scripts/gas-refund-program/transactions-validation/validateTransactions.ts

```diff
@@ -0,0 +1,196 @@
+    // scan transactions in batch sorted by timestamp and hash to guarantee stability
+    const transactionsSlice = await GasRefundTransaction.findAll({
+      where: {
+        epoch: {
+          [Op.gte]: startEpochForTxValidation,
+        },
+      },
+      order: ['timestamp', 'hash'],
+      limit: pageSize,
+      offset,
+      raw: true,
+      attributes: [
+        'id',
+        'chainId',
+        'epoch',
+        'hash',
+        'address',
+        'status',
+        'totalStakeAmountPSP',
+        'gasUsedChainCurrency',
+        'pspChainCurrency',
+        'pspUsd',
+        'refundedAmountUSD',
+        'refundedAmountPSP',
+      ],
+    });
+
+    if (!transactionsSlice.length) break;
+
+    offset += pageSize;
+
+    const updatedTransactions = [];
+
+    for (const tx of transactionsSlice) {
+      let {
+        address,
+        status,
+        totalStakeAmountPSP,
+        gasUsedChainCurrency,
+        pspChainCurrency,
+        pspUsd,
+      } = tx;
+      const refundPercentage = getRefundPercent(totalStakeAmountPSP);
+
+      assert(refundPercentage, 'refundPercentage should be defined and > 0');
+
+      // recompute refundedAmountPSP/refundedAmountUSD as logic alters those values as we reach limits
+      const refundedAmountPSP = new BigNumber(gasUsedChainCurrency)
+        .dividedBy(pspChainCurrency)
+        .multipliedBy(refundPercentage)
+        .decimalPlaces(0);
+
+      const refundedAmountUSD = refundedAmountPSP
+        .multipliedBy(pspUsd)
+        .dividedBy(10 ** 18); // psp decimals always encoded in 18decimals
+
+      let cappedRefundedAmountPSP;
+      let cappedRefundedAmountUSD;
+
+      const isGlobalLimitReached = guardian.isMaxPSPGlobalBudgetSpent();
+      const isPerAccountLimitReached =
+        guardian.isAccountUSDBudgetSpent(address);
+
+      if (isGlobalLimitReached || isPerAccountLimitReached) {
+        status = TransactionStatus.REJECTED;
+      } else {
+        status = TransactionStatus.VALIDATED;
+
+        if (
+          guardian
+            .totalRefundedAmountUSD(address)
+            .plus(refundedAmountUSD)
+            .isGreaterThan(MAX_USD_ADDRESS_BUDGET)
+        ) {
+          cappedRefundedAmountUSD = MAX_USD_ADDRESS_BUDGET.minus(
+            guardian.totalRefundedAmountUSD(address),
+          );
+
+          assert(
+            cappedRefundedAmountUSD.isGreaterThanOrEqualTo(0),
+            'Logic Error: quantity cannot be negative, this would mean we priorly refunded more than max',
+          );
+
+          cappedRefundedAmountPSP = cappedRefundedAmountUSD
+            .dividedBy(pspUsd)
+            .multipliedBy(10 ** 18)
+            .decimalPlaces(0);
+        }
+
+        if (
+          guardian.state.totalPSPRefunded
+            .plus(cappedRefundedAmountPSP || refundedAmountPSP)
+            .isGreaterThan(MAX_PSP_GLOBAL_BUDGET)
+        ) {
+          // Note: updating refundedAmountUSD does not matter if global budget limit is reached
```

Comment:
it only matters for two scenarios.
debugging: the stored refundedUSD amount would be innaccurate - doesn't affect the outcome (distribution numbers).
budget change: if the program budget was subsequentially raised after it was stopped due to reaching the limit (eg restarted later with raised upper limit), then the tx that crossed the original limit would have the wrong usd refunded amount which could then get loaded in to the guardian. (it would be higher than it was) and so that user would have less of their address limit available. Ultimately this is an edge edge case and prob if the GRP finished and was restarted then we could recrunch all data.